cmd/mailbox-controller/controller.go

```diff
@@ -0,0 +1,263 @@
+		syncTargetIndexer:         syncTargetClusterInformer.GetIndexer(),
+		workspaceScopedInformer:   workspacesInformer,
+		workspaceScopedLister:     workspaceScopedPreInformer.Lister(),
+		workspaceScopedClient:     workspaceScopedClient,
+		queue:                     workqueue.NewNamedRateLimitingQueue(workqueue.DefaultControllerRateLimiter(), "mailbox-controller"),
+	}
+
+	syncTargetClusterInformer.AddEventHandler(ctl)
+	workspacesInformer.AddEventHandler(ctl)
+	return ctl
+}
+
+// Run animates the controller, finishing and returning when the context of
+// the controller is done.
+// Call this after the informers have been started.
+func (ctl *mbCtl) Run(concurrency int) {
+	ctx := ctl.context
+	logger := klog.FromContext(ctx)
+	doneCh := ctx.Done()
+	if !cache.WaitForNamedCacheSync("mailbox-controller", doneCh, ctl.syncTargetClusterInformer.HasSynced, ctl.workspaceScopedInformer.HasSynced) {
+		logger.Error(nil, "Informer syncs not achieved")
+		return
+	}
+	logger.V(1).Info("Informers synced")
+	for worker := 0; worker < concurrency; worker++ {
+		go ctl.syncLoop(ctx, worker)
+	}
+	<-doneCh
+}
+
+func (ctl *mbCtl) OnAdd(obj any) {
+	logger := klog.FromContext(ctl.context)
+	logger.V(4).Info("Observed add", "obj", obj)
+	ctl.enqueue(obj)
+}
+
+func (ctl *mbCtl) OnUpdate(oldObj, newObj any) {
+	logger := klog.FromContext(ctl.context)
+	logger.V(4).Info("Observed update", "oldObj", oldObj, "newObj", newObj)
+	if newObj != nil {
+		ctl.enqueue(newObj)
+	} else if oldObj != nil {
+		ctl.enqueue(oldObj)
+	}
+}
+func (ctl *mbCtl) OnDelete(obj any) {
+	logger := klog.FromContext(ctl.context)
+	logger.V(4).Info("Observed delete", "obj", obj)
+	ctl.enqueue(obj)
+}
+
+func (ctl *mbCtl) enqueue(obj any) {
+	logger := klog.FromContext(ctl.context)
+	switch typed := obj.(type) {
+	case *tenancyv1alpha1.Workspace:
+		logger.V(4).Info("Enqueuing reference due to workspace", "wsName", typed.Name)
+		ctl.queue.Add(typed.Name)
+	case *workloadv1alpha1.SyncTarget:
+		mbwsName := mbwsNameOfSynctarget(typed)
+		logger.V(4).Info("Enqueuing reference due to SyncTarget", "wsName", mbwsName, "syncTargetName", typed.Name)
+		ctl.queue.Add(mbwsName)
+	default:
+		logger.Error(nil, "Notified of object of unexpected type", "object", obj, "type", fmt.Sprintf("%T", obj))
+	}
+}
+
+func (ctl *mbCtl) syncLoop(ctx context.Context, worker int) {
+	doneCh := ctx.Done()
+	logger := klog.FromContext(ctx)
+	logger = logger.WithValues("worker", worker)
+	ctx = klog.NewContext(ctx, logger)
+	logger.V(4).Info("SyncLoop start")
+	for {
+		select {
+		case <-doneCh:
+			logger.V(2).Info("SyncLoop done")
+			return
+		default:
+			ref, shutdown := ctl.queue.Get()
+			if shutdown {
+				logger.V(2).Info("Queue shutdown")
+				return
+			}
+			ctl.sync1(ctx, ref)
+		}
+	}
+}
+
+func (ctl *mbCtl) sync1(ctx context.Context, ref any) {
+	defer ctl.queue.Done(ref)
+	logger := klog.FromContext(ctx)
+	logger.V(4).Info("Dequeued reference", "ref", ref)
+	retry := ctl.sync(ctx, ref)
+	if retry {
```

Comment:
Is there a limit on retries ?

Reply:
Deliberately no.  Giving up would mean potentially not eventually recovering from a transient problem.